File: android/app/src/main/java/com/gesture/controller/gesture/ProximityDetector.kt
package com.gesture.controller.gesture

class ProximityDetector(
    private val onWakeDetected: () -> Unit,
    private val onSleepDetected: () -> Unit
) {
    private var lastProximityTime = 0L
    private val sleepThreshold = 3000L // 3 seconds of no hand to sleep
    private var isCurrentlyAwake = false

    fun checkProximity(landmarks: List<NormalizedLandmark>): Boolean {
        if (landmarks.isEmpty()) return false
        val currentTime = System.currentTimeMillis()
        lastProximityTime = currentTime
        if (!isCurrentlyAwake) {
            isCurrentlyAwake = true
            onWakeDetected()
        }
        return true
    }

    fun checkSleep() {
        if (isCurrentlyAwake && System.currentTimeMillis() - lastProximityTime > sleepThreshold) {
            isCurrentlyAwake = false
            onSleepDetected()
        }
    }

    fun reset() {
        lastProximityTime = 0L
        isCurrentlyAwake = false
    }
}

File: src/gesture/ProximityDetector.ts
import { NormalizedLandmark } from '../types';

export class ProximityDetector {
  private lastProximityTime = 0;
  private readonly SLEEP_THRESHOLD = 3000; // 3 seconds of no hand to sleep
  private isCurrentlyAwake = false;

  constructor(
    private onWakeDetected: () => void,
    private onSleepDetected: () => void
  ) {}

  checkProximity(landmarks: NormalizedLandmark[]): boolean {
    if (!landmarks || landmarks.length === 0) return false;
    const currentTime = Date.now();
    this.lastProximityTime = currentTime;
    if (!this.isCurrentlyAwake) {
      this.isCurrentlyAwake = true;
      this.onWakeDetected();
    }
    return true;
  }

  checkSleep(): void {
    if (this.isCurrentlyAwake && Date.now() - this.lastProximityTime > this.SLEEP_THRESHOLD) {
      this.isCurrentlyAwake = false;
      this.onSleepDetected();
    }
  }

  reset(): void {
    this.lastProximityTime = 0;
    this.isCurrentlyAwake = false;
  }
}
